modules/programs/fff.nix

```diff
@@ -0,0 +1,582 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.fff;
+
+in {
+  meta.maintainers = [ maintainers.gigahawk ];
+
+  options.programs.fff = {
+    enable = mkEnableOption "A simple file manager written in bash";
+
+    cdOnExit = {
+      enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Change directory to current folder on exit
+        '';
+      };
+
+      name = mkOption {
+        type = types.str;
+        default = "f";
+        description = ''
+          Name of function used to call fff.
+        '';
+      };
+
+      commands = mkOption {
+        type = types.str;
+        default = ''
+          # cd to exited folder for bash and zsh
+          fff "$@"
+          cd "$(cat "''${XDG_CACHE_HOME:=''${HOME}/.cache}/fff/.fff_d")" 
+        '';
+        example = "";
+        description = ''
+          Bash/Zsh commands used to start fff with cdOnExit enabled.
+        '';
+      };
+
+      fishCommands = mkOption {
+        type = types.str;
+        default = ''
+          fff $argv
+          cd (cat ${cfg.cdOnExit.helperFile})
+        '';
+        description = ''
+          Fish commands used to start fff with cdOnExit enabled.
+        '';
+      };
+
+      helperFile = mkOption {
```

Comment:
Seems like an unnecessary option, it can be removed.

Reply:
I don't imagine a lot of people needing to change this from the default, but I also don't know if I would deem it unnecessary.
I figure someone might want to configure this if it's listed as an option upstream.

Reply:
In general it is best to only expose explicit options for those settings that are absolutely necessary for getting a functional setup. Any other options that the user would like to set should be settable through an `extraConfig` option. See [3.2.3. Add only valuable options](https://rycee.gitlab.io/home-manager/index.html#sec-guidelines-valuable-options).

In this case I imagine `extraConfig` having the type `attrsOf (oneOf [bool int str])` and for convenience it could automatically prepend the `FFF_` part and upper case the attribute names. So, this option could be set by a user through
```
extraConfig = {
  cd_file = "${config.home.homeDirectory}/custom/path";
}
```

Reply:
Actually, having looked through the options now I think you could move to a mostly [RFC 42](https://github.com/NixOS/rfcs/pull/42) style configuration by just having
```
settings = mkOption {
  type = with types; attrsOf (oneOf [bool int str]);
  apply =
    let
      normalizeName = n: if hasPrefix "FFF_" n then n else "FFF_" + toUpper n;
    in mapAttrs' (n: v: { name = normalizeName n; value = v; });
  default = {};
  example = {
    trash_cmd = "trash";
  };
  description = ''
    … see https://github.com/dylanaraps/fff#customization for available options …
    … an option FFF_ABC can here optionally be written abc …
  '';
}

keyBindings = mkOption {
  type = with types; attrsOf str;
  # Rest much like settings above but using FFF_KEY_ instead of FFF_
  …
}
```

Then in the `config` section have something like
```
home.sessionVariables =
  {
    FFF_CD_FILE = "${config.xdg.cacheHome}/fff/.fff_d";
    # Any other options that must be set.
  }
  // cfg.keyBindings
  // cfg.settings;
```